aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -139,7 +140,11 @@ private InputStream readRange(String range) {
 
     S3RequestUtil.configureEncryption(s3FileIOProperties, requestBuilder);
 
-    return s3.getObject(requestBuilder.build(), ResponseTransformer.toInputStream());
+    stream =
+        RetryableInputStream.builderFor(
+                () -> s3.getObject(requestBuilder.build(), ResponseTransformer.toInputStream()))
```

Comment:
any failures that occur on the retry during the getObject request (not the stream read) should just be handled by the SDK. Don't think we need to add anything custom for that since the S3 client is already pluggable.

Reply:
I feel like there's an issue here.  When the stream gets recreated, the stream will reset back to the original position and we continue from there as if we're at the right place in the stream.

The `pos` won't reflect the new position of the stream, if I'm reading this correctly.  I would think the following retry would need to start from `next` to reflect where the next read should start.  There's a small problem with the single byte read method because we increment the positions prior to read, so that would likely need to be adjusted to be after the read like in the range read method.

Reply:
I'll double check this, I think you're right although I'm not sure why the first test case wouldn't surface that (since the content would be different). We may need to seek properly on the input stream initialization during retry.

Reply:
I updated this, the RetryableInputStream offers two builder APIs, one for specifying just a new stream initialization and one for a stream initialization plus a position supplier. The stream initialization function takes in a position (the position can be null to handle the range based requests since for range reads with explicit begin/end we don't care about the current position in the stream). cc @danielcweeks

Reply:
Discussed with @danielcweeks , for the range read cases since we're using IOUtil readFully/readRemaining which will read the range in a buffered manner. On retries we would read from the beginning position but the internal stream tracking in readFully/readRemaining would not reset to the right position in the buffer to read so that's still an issue.

What we can do is to just not retry the RangeReadable methods for now since they're not actually exercised anywhere.  Down the line, we could just use FailSafe and retry on the whole method.